lib/hllcounter.cc

```diff
@@ -269,36 +187,33 @@ HLLCounter::HLLCounter(double error_rate, WordLength ksize)
         throw InvalidValue("Please set error rate to a value "
                            "greater than zero");
     }
-    int p = ceil(log2(pow(1.04 / error_rate, 2)));
-    this->init(p, ksize);
+    int new_p = ceil(log2(pow(1.04 / error_rate, 2)));
+    init(1 << new_p, ksize);
 }
 
-HLLCounter::HLLCounter(int p, WordLength ksize)
+HLLCounter::HLLCounter(int nc, WordLength ksize)
 {
-    this->init(p, ksize);
+    init(nc, ksize);
 }
 
-void HLLCounter::init(int p, WordLength ksize)
+void HLLCounter::init(int nc, WordLength ksize)
 {
-    this->alpha = calc_alpha(p);
-    this->p = p;
-    this->_ksize = ksize;
-    this->m = 1 << p;
-    std::vector<int> M(this->m, 0.0);
-    this->M = M;
-
-    init_raw_estimate_data();
-    init_bias_data();
+    alpha = calc_alpha(nc);
+    ncounters_log2 = floor(log2(nc));
+    ncounters = 1 << ncounters_log2;
+    _ksize = ksize;
+    std::vector<uint8_t> newc(ncounters, 0);
+    counters = newc;
 }
 
 double HLLCounter::get_erate()
 {
-    return 1.04 / sqrt(this->m);
+    return 1.04 / sqrt(ncounters);
 }
 
 void HLLCounter::set_erate(double error_rate)
 {
-    if (count(this->M.begin(), this->M.end(), 0) != this->m) {
+    if (count(begin(counters), end(counters), 0) != ncounters) {
```

Comment:
Do you want to move this to a method with a name like `_all_counters_zero()`? It took me a moment to work out what it was doing and the code is reused below.